pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
 	return cb.WithExpectedErrs(expectedFunc)
 }
 
-var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	cfgName, cfg := cb.cfg.findIterConfig(dir)
+	if cfg == nil {
+		return cb.Bucket.Iter(ctx, dir, f)
+	}
+
+	cb.operationRequests.WithLabelValues(opIter, cfgName).Inc()
+
+	key := cachingKeyIter(dir)
+	data := cfg.cache.Fetch(ctx, []string{key})
+	if data[key] != nil {
+		list, err := cfg.codec.Decode(data[key])
+		if err == nil {
+			cb.operationHits.WithLabelValues(opIter, cfgName).Inc()
+			for _, n := range list {
+				if err := f(n); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		level.Warn(cb.logger).Log("msg", "failed to decode cached Iter result", "err", err)
+	}
+
+	// Iteration can take a while (esp. since it calls function), and iterTTL is generally low.
+	// We will compute TTL based time when iteration started.
+	iterTime := time.Now()
+	var list []string
+	err := cb.Bucket.Iter(ctx, dir, func(s string) error {
+		list = append(list, s)
+		return f(s)
+	})
+
+	remainingTTL := cfg.ttl - time.Since(iterTime)
+	if err == nil && remainingTTL > 0 {
```

Comment:
`remainingTTL > 0` 

I might think this as error case if users defines ttl smaller than actual iteration time (:

Reply:
I would log warning at least + metric.

Thanks for caring for this BTW! and spotting this special case :muscle:

Reply:
I use the same `ttl > 0` pattern in all methods. If we add logging + metric here, then I think it should go to all of them.

Reply:
Agree, but would be nice to log+metric, but we can do as follow up